src/plugins/factoids/storage.internal.js

```diff
@@ -0,0 +1,171 @@
+/* eslint-disable no-underscore-dangle */
+const fs = require('fs');
+const path = require('path');
+const { promisify, inspect } = require('util');
+
+const readFile = promisify(fs.readFile);
+const writeFile = promisify(fs.writeFile);
+
+const FILE_PATH = path.join(__dirname, 'factoids.json');
+
+function toKey(key) {
+  return (key && key.toLowerCase()) || null;
+}
+
+class Store {
+  constructor() {
+    this._items = new Map();
+    this._loaded = false;
+    this._readPromise = null;
+    this._writePromise = null;
+    this._dirty = false;
+  }
+
+  _replaceAll(items) {
+    this._items = items;
+  }
+
+  set(key, entry) {
+    this._items.set(key.toLowerCase(), entry);
+    this._dirty = true;
+    return this;
+  }
+
+  getText(key) {
+    const MAX_ALIAS_DEPTH = 5;
+    let cursor = { type: 'alias', value: toKey(key) };
+
+    for (
+      let i = 0;
+      i < MAX_ALIAS_DEPTH && cursor && cursor.type === 'alias';
+      i += 1
+    ) {
```

Comment:
Please recommend a better way to do this while loop with an iteration limit.

Reply:
would it be possible to resolve aliases at save-time, so read is always O(1)?

Reply:
You can update the thing the alias points to. 
```
me: !learn alias foo = bar
me: !learn bar = new value
me: !foo
jelllbot: new value
```

We could do depth and circular reference checking at `!learn` time, though.

Reply:
If we wanted O(1) lookup, we'd need to maintain a mapping of aliases to factoids in memory, which is extra synchronization work. I'm not worried about the cost of resolving a few aliases on factoid usage; especially considering the cost is only paid when it is an actual alias.

Storing aliases on the thing they point to is much more expensive, and the data model would allow creating the same alias twice without special checks for it existing anywhere else, as well as !learn conflicting with an alias. The way both the previous and new file format deal with aliases seems to solve all of these problems fairly well.

I think I'd like to prevent a dangling alias simply because it would typically be a mistake, and make it easier to end up with a circular alias.

Limiting alias depth also limits complexity (to users of the bot), and I'm not convinced we even need more than one level of aliases (i.e. it might be fine to forbid an alias to an alias).